### PR TITLE
[InstCombine] Retain debug information on store to null instruction

### PR DESCRIPTION
**Pull request by quic-asaravan**
Call InsertNewInstWith() instead of InsertNewInstBefore() when creating "store to null" instruction

### CODE REVIEW

**Comment by quic-asaravan** [comment]
Debug information not attached to the "store to null" instruction inserted in InstCombinePass to mark unreachable code.

cpp code which generates store to null after instcombinePass (test.cpp):

```
void *def;
#define s_node &def

typedef enum {
A,
B,
C,
D,
E
} enum_type;

typedef struct {
    void *field1;
} struct_1;

typedef struct {
  struct_1 *field1;
  void *val;
} struct_2;

typedef struct s {
  struct s *prev, *next;
  struct_2 *r;
  void *val;
} struct_3;

void foo();
void bar();
void foo1();
void bar1(struct_2 *val);

static unsigned int baz( struct_3 *L )
{
  return ( L->val != L->r->field1->field1 );
}

static enum_type baz1( struct_2 *r )
{
  if ( !r )
  {
    return A;
  }
  if ( r->field1 )
  {
    return B;
  }
  if ( s_node == r->val )
  {
    return C;
  }
  if ( r->val )
  {
    return D;
  }

  return E;
}

static unsigned int helper( struct_2     *r,
                                                     enum_type type )
{
  return ( type == baz1( r ) );
}

void F( struct_3 *pool, struct_3 *link);

struct {
    struct_3 *r;
    struct_3 p;
} struct_4;

#define NULL nullptr
#define VALIDATE(x) 

void foo( void )
{
  struct_3 *val;
  struct_3 *next_val;
  struct_3  sen;
  bar();
  val = struct_4.r; 
  sen.next = val;
  sen.prev = NULL;

  if ( val ) 
  { 
    VALIDATE( NULL == val->prev );
    val->prev = &sen;
  }
  
  /* First, remove all the aliases/placeholders */
  while ( val )
  {
    next_val = val->next;
    if ( helper( val->r, B ) ||
         baz( val ) )
    {
      val->prev->next = next_val;
      if ( next_val )
      {
        next_val->prev = val->prev;
      }
      
      F( &struct_4.p, val );
    }
    val = next_val;
  }

  /* Now remove all the remaining links */
  val = sen.next;
  while ( val )
  {
    next_val = val->next;

    val->prev->next = next_val;
    if ( next_val )
    {
      next_val->prev = val->prev;
    }

    bar1( val->r );
    F( &struct_4.p, val );
    
    val = next_val;
  }
  struct_4.r = sen.next;
  VALIDATE( NULL == struct_4.r );
  foo1();
}

```

The above code is compiled with clang (compiled at SHA - https://github.com/llvm/llvm-project/commit/10451ded6d7ef799569652dfce81653d37f167b5)

`clang++ -g -O2 -mllvm -print-after-all -mllvm -filter-print-funcs="_Z3foov" test.cpp -c`

```
.
.
.
*** IR Dump After InstCombinePass on _Z3foov ***
; Function Attrs: mustprogress
define dso_local void @_Z3foov() local_unnamed_addr #0 !dbg !49 {
.
.
lor.lhs.false:                                    ; preds = %_ZL6helperP8struct_29enum_type.exit
  call void @llvm.dbg.value(metadata ptr %val.058, metadata !106, metadata !DIExpression()), !dbg !113
  store i1 true, ptr poison, align 1            **<======== Debug Information missing ========>**
  br i1 poison, label %if.end15, label %if.then7, !dbg !114
.
.

```
Any specific reason not to have debug information attached to "store to null" instruction?

**Comment by adrian-prantl** [comment]
If none of the cases outlined in https://llvm.org/docs/HowToUpdateDebugInfo.html apply, and there is no comment in the existing code, chances are good that this was just an oversight in the existing implementation. Is the store in the same basic block as indicated by the source location?

**Comment by quic-asaravan** [comment]
> This'll need test coverage (a fully reduced IR test case)
> 
> And it'd be worth having an explanation of what instructions are the input to this instruction (to know what location it should have) and whether they're in the same basic block or not (it's OK to preserve the location if we aren't making locations reachable that were not reachable in the original program): https://llvm.org/docs/HowToUpdateDebugInfo.html

I tried my best to reduce the test case. Find attached the lit test.
[test.txt](https://github.com/llvm/llvm-project/files/15008571/test.txt)

**Comment by quic-asaravan** [comment]
@dwblaikie @dtcxzyw Can you please review the attached test case and let me know?

**Comment by felipepiovezan** [comment]
> Reference

†

**Comment by felipepiovezan** [comment]
My bad, I misclicked this.

**Comment by felipepiovezan** [comment]
> @dwblaikie @dtcxzyw Can you please review the attached test case and let me know?

I had a look at the test case, but it's still pretty big, with lots of metadata, and the single check statement seems like it could match against something else. My suggestion: try reducing the test case using [llvm-reduce](https://llvm.org/docs/CommandGuide/llvm-reduce.html): you can write a script that uses the compiler without the fix and checks if the bug is still there on every reduction/

**Comment by dwblaikie** [comment]
The Sony folks tend to be the ones most familiar with debug locations at the moment who might have some ideas about how to write more targeted tests here. @OCHyams @jmorse @SLTozer 

Not sure what techniques they usually use, but for middle end optimizations, I'd have thought maybe a constructive approach (building up a test case from scratch, more or less) might be more suitable than trying to reduce one - well, maybe reducing at the IR level (without debug info) then adding in enough debug info to tickle the merging problem by hand.

But probably don't run with that ^ suggestion until you hear back from the Sony folks,  as they've probably got better ideas that are less likely to be a huge time sink/send you the wrong way.

**Comment by OCHyams** [comment]
Reducing `test.txt` (attached earlier) using `llvm-reduce` gets me:
```
source_filename = "test.ll"
target triple = "hexagon"

define void @_Z3foov() {
entry:
  tail call void null(), !dbg !11
  ret void
}

!llvm.dbg.cu = !{!0}
!llvm.module.flags = !{!10}

!0 = distinct !DICompileUnit(language: DW_LANG_C_plus_plus_14, file: !1, producer: "QuIC LLVM Hexagon Clang version 8.7-alpha1 Engineering Release: hexagon-clang-87-43126", isOptimized: true, runtimeVersion: 0, emissionKind: FullDebug, enums: !2, globals: !3, splitDebugInlining: false, debugInfoForProfiling: true, nameTableKind: None)
!1 = !DIFile(filename: "test_sym_mod.cpp", directory: "/local/mnt/workspace/asaravan/QTOOL-111650/test_bugpoint")
!2 = !{}
!3 = !{!4, !7}
!4 = !DIGlobalVariableExpression(var: !5, expr: !DIExpression())
!5 = distinct !DIGlobalVariable(name: "def", scope: !0, file: !1, line: 1, type: !6, isLocal: false, isDefinition: true)
!6 = !DIDerivedType(tag: DW_TAG_pointer_type, baseType: null, size: 32)
!7 = !DIGlobalVariableExpression(var: !8, expr: !DIExpression())
!8 = distinct !DIGlobalVariable(name: "struct_4", scope: !0, file: !1, line: 70, type: !9, isLocal: true, isDefinition: true)
!9 = distinct !DICompositeType(tag: DW_TAG_structure_type, file: !1, line: 67, size: 160, flags: DIFlagTypePassByValue, elements: !2)
!10 = !{i32 2, !"Debug Info Version", i32 3}
!11 = !DILocation(line: 83, column: 3, scope: !12)
!12 = distinct !DISubprogram(name: "foo", linkageName: "_Z3foov", scope: !1, file: !1, line: 77, type: !13, scopeLine: 78, flags: DIFlagPrototyped | DIFlagAllCallsDescribed, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !0, retainedNodes: !2)
!13 = !DISubroutineType(types: !14)
!14 = !{null}
```

`opt -passes=instcombine test.ll -S`
```
define void @_Z3foov() {
entry:
  store i1 true, ptr poison, align 1  ; missing !dbg
  ret void
}

... metadata ...
```

That only tests one of the 6 call sites of `CreateNonTerminatorUnreachable`. @dwblaikie is that enough coverage?

All the call sites of `CreateNonTerminatorUnreachable` pass the instruction that's being replaced/marked as unreachable in as the `InsertAt` argument. That is to say that this patch does the right thing currently - it applies the source location from an instruction that is being replaced to the one replacing it. However, there's no guarantee that a future usage of `CreateNonTerminatorUnreachable` passes in the to-be-replaced instruction as `InsertAt`, but it seems unlikely that wouldn't be the case.

Overall SGTM once the test has been added.

**Comment by quic-asaravan** [comment]
@OCHyams Thanks for the reduced test case. @dwblaikie Can you please let us know if this is enough coverage? If yes, can we merge it ?

**Comment by felipepiovezan** [comment]
> @OCHyams Thanks for the reduced test case. @dwblaikie Can you please let us know if this is enough coverage? If yes, can we merge it ?

Please update the PR so we can see what the final test will look like, with the relevant `CHECK` lines :)
But yeah, in general the test above is much better than the original one

**Comment by dwblaikie** [comment]
Yeah, always a quesiton about how far up to go to get adequate coverage - at least one caller's enough for me when I'm not looking too closely, like now/here :)

File: llvm/test/Transforms/InstCombine/StoreToNull-DbgCheck.ll [hunk 1]
```diff
@@ -0,0 +1,30 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
+; Verify whether debug information is attached to the store to null/undef instruction
+; !dbg should be present at the end of "store to null" instruction
+; RUN: opt -S -passes='instcombine' < %s | FileCheck %s > /dev/null
+
+source_filename = "test.ll"
+target triple = "hexagon"
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/StoreToNull-DbgCheck.ll:7]: Remove source_filename and triple.

File: llvm/test/Transforms/InstCombine/StoreToNull-DbgCheck.ll [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+define void @_Z3foov() {
+; CHECK-LABEL: define void @_Z3foov() {
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    store i1 true, ptr poison, align 1, !dbg
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/StoreToNull-DbgCheck.ll:12]: It looks like you modified the UTC output by hand. Do not modify the UTC output.

File: llvm/test/Transforms/InstCombine/StoreToNull-DbgCheck.ll [hunk 3]
```diff
@@ -0,0 +1,27 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
+; Verify whether debug information is attached to the store to null/undef instruction
+; !dbg should be present at the end of "store to null" instruction
+; RUN: opt -S -passes='instcombine' < %s | FileCheck %s > /dev/null
+
+define void @_Z3foov() {
+; CHECK-LABEL: define void @_Z3foov() {
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    store i1 true, ptr poison, align 1
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/StoreToNull-DbgCheck.ll:9]: This is now lacking the `, !dbg` attachment entirely -- did you use the correct build?

File: llvm/test/Transforms/InstCombine/StoreToNull-DbgCheck.ll (hunk 2 above)
Comment (nikic) [reply llvm/test/Transforms/InstCombine/StoreToNull-DbgCheck.ll:9]: And now you're back to manually modifying the CHECK lines. You need to run update_test_checks.py with the correct opt binary. The call will look something like `llvm/utils/update_test_checks.py --opt-binary build/bin/opt llvm/test/Transforms/InstCombine/StoreToNull-DbgCheck.ll` (obviously after you have built LLVM with your changes...) If you do that, then this should produce something like `, !dbg [[DBG1:![0-9]+]]`.

Comment (quic-asaravan) [reply llvm/test/Transforms/InstCombine/StoreToNull-DbgCheck.ll]: @nikic I have updated the lit test. Can you please review it?

**Review by nikic — APPROVED** [review]
LGTM

**Comment by quic-asaravan** [comment]
@nikic Can you also please merge it as I don't have write access.

**Pull request merged**